frontends/p4/simplify.cpp

```diff
@@ -92,16 +93,19 @@ const IR::Node* DoSimplifyControlFlow::postorder(IR::EmptyStatement* statement)
 const IR::Node* DoSimplifyControlFlow::postorder(IR::SwitchStatement* statement)  {
     LOG3("Visiting " << dbp(getOriginal()));
     if (statement->cases.empty()) {
-        // The P4_16 spec prohibits expressions other than table application as
-        // switch conditions.  The parser should have rejected programs for
-        // which this is not the case.
-        BUG_CHECK(statement->expression->is<IR::Member>(),
-                  "%1%: expected a Member", statement->expression);
-        auto expr = statement->expression->to<IR::Member>();
-        BUG_CHECK(expr->expr->is<IR::MethodCallExpression>(),
-                  "%1%: expected a table invocation", expr->expr);
-        auto mce = expr->expr->to<IR::MethodCallExpression>();
-        return new IR::MethodCallStatement(mce->srcInfo, mce);
+        // If this is a table application remove the switch altogether but keep
```

Comment:
Do we have a test for this?

Reply:
Also another test for a function call with side-effects could be useful.

Reply:
issue3619-1.p4 is testing the side-effect.

Reply:
But that function call is side-effect free. Also this does not include a table.apply(), right?

Reply:
This pass does not know that the function has no side effects. All tests with table applications would qualify for the second case.

Reply:
```
if (SideEffects::check(statement->expression, this, refMap, typeMap))
    // This can happen if this pass is run before SideEffectOrdering.
    return statement;
```
I am trying to understand this segment. If the function call has side effects, ideally only the method call statement should remain, right? I do not think we have a test for this. 

Similarly, 
`if (TableApplySolver::isActionRun(statement->expression, refMap, typeMap)) {`
should check whether there is an action_run call and then only leave the table apply statement. 

I assume we have tests already for this case? Since this was the default behavior.

Reply:
If the argument of the switch may have side-effects then we leave it unchaged. This pass is invoked again later, after SideEffectOrdering has removed calls with side effects from switch statements (all except table.apply().action_run), and that case will never occur again.